File: backend/model-domain/src/main/java/fi/hel/allu/model/domain/DefaultAttachmentInfo.java
package fi.hel.allu.model.domain;

import fi.hel.allu.common.types.ApplicationType;
import fi.hel.allu.common.types.AttachmentType;

import java.time.ZonedDateTime;
import java.util.List;

/**
 * Default attachment info adds default attachment specific information to the regular attachment.
 */
public class DefaultAttachmentInfo extends AttachmentInfo {
  private Integer defaultAttachmentId;
  private List<ApplicationType> applicationTypes;
  private Integer fixedLocationAreaId;

  public DefaultAttachmentInfo() {
  }

  public DefaultAttachmentInfo(
      Integer id,
      Integer userId,
      AttachmentType type,
      String name,
      String description,
      Long size,
      ZonedDateTime creationTime,
      Integer defaultAttachmentId,
      List<ApplicationType> applicationTypes,
      Integer fixedLocationAreaId) {
    super(id, userId, type, name, description, size, creationTime);
    this.defaultAttachmentId = defaultAttachmentId;
    this.applicationTypes = applicationTypes;
    this.fixedLocationAreaId = fixedLocationAreaId;
  }

  /**
   * Returns the database id of default attachment (different than the attachment id of super class).
   *
   * @return  the database id of default attachment (different than the attachment id of super class).
   */
  public Integer getDefaultAttachmentId() {
    return defaultAttachmentId;
  }

  public void setDefaultAttachmentId(Integer defaultAttachmentId) {
    this.defaultAttachmentId = defaultAttachmentId;
  }

  /**
   * Returns list of application types using the default attachment.
   *
   * @return  list of application types using the default attachment. May be <code>null</code>.
   */
  public List<ApplicationType> getApplicationTypes() {
    return applicationTypes;
  }

  public void setApplicationTypes(List<ApplicationType> applicationTypes) {
    this.applicationTypes = applicationTypes;
  }

  /**
   * Returns id of the fixed location area related to the default attachment.
   *
   * @return  id of the fixed location area related to the default attachment. May be <code>null</code>.
   */
  public Integer getFixedLocationAreaId() {
    return fixedLocationAreaId;
  }

  public void setFixedLocationAreaId(Integer fixedLocationAreaId) {
    this.fixedLocationAreaId = fixedLocationAreaId;
  }
}

File: frontend/src/app/model/application/attachment/default-attachment-info.ts
import {AttachmentInfo} from './attachment-info';
import {TimeUtil} from '../../../util/time.util';
import {DefaultAttachmentInfoForm} from './default-attachment-info-form';
import {AttachmentType} from '@model/application/attachment/attachment-type';

export class DefaultAttachmentInfo extends AttachmentInfo {
  constructor(
    public id?: number,
    public type?: AttachmentType,
    public mimeType?: string,
    public name?: string,
    public description?: string,
    public size?: number,
    public creationTime?: Date,
    public decisionAttachment?: boolean,
    public handlerName?: string,
    public file?:  Blob | File,
    public defaultAttachmentId?: number,
    public applicationTypes?: Array<string>,
    public fixedLocationId?: number
  ) {
    super(id, type, mimeType, name, description, size, creationTime, decisionAttachment, handlerName, file);
  }

  static fromForm(form: DefaultAttachmentInfoForm): DefaultAttachmentInfo {
    return new DefaultAttachmentInfo(
      form.id,
      form.type,
      form.mimeType,
      form.name,
      form.description,
      form.size,
      TimeUtil.getDateFromUi(form.creationTime),
      true,
      form.handlerName,
      form.file,
      form.defaultAttachmentId,
      form.applicationTypes,
      form.fixedLocationId
    );
  }

  static toForm(attachmentInfo: DefaultAttachmentInfo): DefaultAttachmentInfoForm {
    return {
      id: attachmentInfo.id,
      type: attachmentInfo.type,
      mimeType: attachmentInfo.mimeType,
      name: attachmentInfo.name,
      description: attachmentInfo.description,
      creationTime: attachmentInfo.uiCreationTime,
      decisionAttachment: attachmentInfo.decisionAttachment,
      handlerName: attachmentInfo.handlerName,
      file: attachmentInfo.file,
      defaultAttachmentId: attachmentInfo.defaultAttachmentId,
      applicationTypes: attachmentInfo.applicationTypes,
      fixedLocationId: attachmentInfo.fixedLocationId
    };
  }
}

